File: demo-web/src/main/java/ru/gnkoshelev/demo/spring_angular/web/service/TagService.java
package ru.gnkoshelev.demo.spring_angular.web.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import ru.gnkoshelev.demo.spring_angular.web.repository.TagRepository;
import ru.gnkoshelev.demo.spring_angular.web.so.api.TagsRequester;
import ru.gnkoshelev.demo.spring_angular.web.so.api.TagsParser;
import ru.gnkoshelev.demo.spring_angular.web.model.Tag;

import java.util.Arrays;
import java.util.List;

/**
 * Created by kgn on 20.09.2015.
 */
@Service
public class TagService {
    @Autowired
    private TagsRequester requester;

    @Autowired
    private TagRepository repository;

    @Autowired
    private TagsParser parser;

    @Value(value = "${suggestedLimit}")
    private Integer limit;

    public Tag get(String name) {
        Tag tag = repository.get(name);
        if (tag != null) {
            return tag;
        }
        String rawTag = requester.getTagInfo(name);
        if (rawTag == null) {
            return null;
        }
        tag = parser.parseTag(name, rawTag);
        return tag != null ? repository.persist(tag) : null;
    }

    public List<String> suggest(String pattern) {
        List<String> suggested = repository.suggest(pattern, limit);
        return suggested;
    }

    public List<Tag> list(String[] names) {
        List<Tag> tags = repository.list(Arrays.asList(names));
        return tags;
    }
}

File: demo-web/src/main/webapp/WEB-INF/client/ts/services/TagService.ts
import Promise = require('commons/Promise');
import Tag = require('model/Tag');

class TagService {
    public http:ng.IHttpService;
    public contextPath:string;
    public q:ng.IQService;
    public url:string;

    public static injection():any[] {
        return ['$http', 'contextPath', '$q', ($http:ng.IHttpService, contextPath:string, $q:ng.IQService) => new TagService($http, contextPath, $q)];
    }

    public constructor($http:ng.IHttpService, contextPath:string, $q:ng.IQService) {
        this.http = $http;
        this.contextPath = contextPath;
        this.q = $q;
        this.url = this.contextPath + "/api/tags";
    }

    public suggest(pattern: string): ng.IPromise<string[]> {
        return Promise.promise(
            this.http.get<string[]>(
                this.url + "/suggest",
                {
                    params: {
                        pattern: pattern
                    }
                }
            ),
            this.q
        );
    }

    public list(names: string[]): ng.IPromise<Tag[]> {
        return Promise.promise(
            this.http.get<Tag[]>(
                this.url + "/list",
                {
                    params: {
                        names: (names != null) ? names.join(' ') : null
                    }
                }),
            this.q)
    }

    public get(name: string): ng.IPromise<Tag> {
        return Promise.promise(
            this.http.get<Tag>(
                this.url + "/get",
                {
                    params: {
                        name: name
                    }
                }
            ),
            this.q);
    }
}

export = TagService;